### PR TITLE
Add support for easily disabling metrics export

### PR DESCRIPTION
**Pull request by bono007**
Add global metrics exporter disable property; set during tests with ability to override via `@AutoConfigureMetrics`.

Closes gh-21616

### CODE REVIEW

File: spring-boot-project/spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/metrics/ExcludeMetricExportersContextCustomizerFactoryTests.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for {@link AutoConfigureMetrics} and
+ * {@link ExcludeMetricExportersContextCustomizerFactory} working together.
+ *
+ * @author Chris Bono
+ */
+class ExcludeMetricExportersContextCustomizerFactoryTests {
```
Comment (bono007) [inline spring-boot-project/spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/metrics/ExcludeMetricExportersContextCustomizerFactoryTests.java:33]: This test and the sibling ones `AutoConfigureMetricsMissing/PresentIntegrationTest` overlap on coverage. If I choose between them I think that the integration test ones are more valuable as they verify against `@SpringBootTest` as the annotated class.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/export/appoptics/AppOpticsMetricsExportAutoConfigurationTests.java [hunk 2]
```diff
@@ -52,7 +52,14 @@ void autoConfiguresItsConfigAndMeterRegistry() {
 	}
 
 	@Test
-	void autoConfigurationCanBeDisabled() {
+	void autoConfigurationCanBeDisabledWithGlobalEnabledProperty() {
```
Comment (bono007) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/export/appoptics/AppOpticsMetricsExportAutoConfigurationTests.java:55]: I will do this same mod to the other export config tests once I know we are good w/ the direction of this proposal.

**Comment by bono007** [comment]
@snicoll here is what I captured along the way while trying different approaches.

## Background

*Metircs auto-configuration topology (wrt AutoConfigureBefore/After of each other)*
```
MetricsAutoConfiguration	
	WavefrontMetricsExportAutoConfiguration (enabled!=false -> WavefrontMeterRegistry)
	DatadogMetricsExportAutoConfiguration (enabled!=false -> DatadogMeterRegistry)
	..
		SimpleMetricsExportAutoConfiguration (COMB-MR, enabled!=false -> SimpleMeterRegistry)
			CompositeMeterRegistryAutoConfiguration	
					- NoOpMeterRegistryConfiguration (COMB-MR -> CompositeMeterRegistry empty)
					- CompositeMeterRegistryConfiguration (multi MRs -> CompositeMeterRegistry w/ MRs) 
				MetricsEndpointAutoConfiguration
```

## Goal
To disable everything between `MetricsAutoConfiguration` and `SimpleMetricsExportAutoConfiguration` by default in tests. With ability to override and re-enable all/individual exporters.

## Approaches
I tried several approaches leveraging `ImportAutoConfiguration` and built-in auto-config exclusion etc. but they all ended w/ non-success or non-starters shortcomings. 

*TLDR;* Jump to approach #5 to see the content of this code proposal.


The following are of the "Don't modify any of the existing exporter auto-configurations" variety:

*Approach #1:*
------------
- @AutoConfigureMetrics as an @ImportAutoConfiguration which maps to list of ACs to use from spring.factories
- AutoConfigurationImportFilter that exclude ACs ending in "MetricsExportAutoConfiguration"
*Status:* Does not work. The ACIF exclusions take precedence over the IAC list.

*Approach #2:* 
------------
- @AutoConfigureMetrics as an @ImportAutoConfiguration which maps to list of ACs to use from spring.factories
- ContextCustomizer that sets the 'spring.autoconfigure.exclude' property to list of ACs from spring.factories (above)
*Status:* Does not work. The exclusion always overrides the inclusions.

*Approach #3:*
------------
- @AutoConfigureMetrics simple marker interface, nothing more
- ContextCustomizer that sets 'spring.autoconfigure.metric-exporters.enabled' based on whether the test is marked as @AutoConfigureMetrics
- AutoConfigurationImportFilter that exclude ACs ending in "MetricsExportAutoConfiguration" unless above property is set (aka marked w/ @AutoConfigureMetrics)

*Status:* works 
- Pros:
  - simple to implement (no change to existing auto-configs)
  - users custom metric exporters will be respected IF named properly
- Cons: 
  - excludes by naming scheme ("*MetricExportAutoConfiguration")
  - binary enablement (all or nothing) - no way to override and add an inclusion/exclusion

*Approach #4:*
------------
- @AutoConfigureMetrics simple marker interface, nothing more
- ContextCustomizer that sets the 'spring.autoconfigure.exclude' property to hardcoded set of metric exporter configs unless the test is marked w/ @AutoConfigureMetrics 

*Status:* works 
- Pros:
  - simple to implement (no change to existing auto-configs)
- Cons: 
  - excludes by naming scheme ("*MetricExportAutoConfiguration") (NOTE: this can be alleviated by reading list from spring.factories)
  - binary enablement (all or nothing)
  - users custom metric exporters will not be respected in exclusion


The following is of the "Do what you need to do to make this work properly" variety:

*Approach #5:*
------------
- @AutoConfigureMetrics simple marker interface, nothing more
- ContextCustomizer that sets new "global" 'management.metrics.export.enabled' based on whether the test is marked as @AutoConfigureMetrics
- @ConditionalOnMetricExportsEnabled that inspects global and specific property to decide match or not

*Status:* works
- Pros:
  - allows all export to be disabled w/ single property but respects specific export overrides
  - can be used outside of test-land
- Cons: 
  - small change to existing auto-configs - but not functional breaking change
  - users custom metric exporters will be respected in exclusion if they copy pattern and use the @ConditionalOnMetricExportsEnabled


## Outcome
New integration test defaults is that only `SimpleMeterRegistry` will be enabled. The `management.metrics.export[.<name>].enabled` properties can be used to adjust that. 

## Next
I have submitted a handful of tests to sanity check the functionality and illustrate the usage of the new annotation. If this is a direction that we agree on then I will add the missing tests and update the docs accordingly in a subsequent commit.

File: spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/metrics/AutoConfigureMetrics.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation that can be applied to a test class to enable auto-configuration for metrics
+ * exporters. By default, all metrics exporters other than the in-memory
+ * {@code SimpleMetricsExportAutoConfiguration} are disabled.
```
Comment (snicoll) [inline spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/metrics/AutoConfigureMetrics.java:29]: I think that could confuse users as it can be read what the default behaviour of adding the annotation does. I'd rather keep it simple and we can handle the rationale in the reference guide rather.

**Comment by bono007** [comment]
@snicoll 
I should have to add the tests and update the comments based on the feedback w/in the next 12 hours.

**Comment by bono007** [comment]
@snicoll - I added the auto-config tests for each exporter and simplified the java doc.

**Comment by bono007** [comment]
Is the failure around `RestAssuredRestDocsAutoConfigurationAdvancedConfigurationIntegrationTests` expected? Looks like its getting a 401 and when i hit the endpoint that its trying to hit I see a form based login screen appear. I don't think anything in this merge request could affect that.

**Comment by snicoll** [comment]
Same. Don't worry about it, I'll investigate as part of polishing this change.

**Comment by snicoll** [comment]
> I don't think anything in this merge request could affect that.

Actually, it does indirectly as the actuator (and it's auto-configuration) is now added to the project. I don't know why that happens but on the other hand Actuator does not seem required. I'll investigate a bit more.

**Comment by bono007** [comment]
Ahh interesting. Sounds like security is kicking in now with the added deps.

**Comment by bono007** [comment]
@snicoll did you get a chance to follow up on the issue? If not, I can take a look in the next 24hrs.

**Comment by snicoll** [comment]
I've removed actuator as it's not required although I might change my mind and write a test with an actual registry implementation to validate it's auto-configuration isn't enabled. I am on PTO the next 2 days and I've already started to polish things quite a bit in a branch. If you want to investigate how to prevent security to kick in the way it does, I am definitely interested. Thanks!

**Comment by bono007** [comment]
Enjoy your PTO! I will look into how to prevent security from kicking in as
it will be good to know, regardless of final direction.

**Comment by bono007** [comment]
I just got around to look into this now. 

1. Yes, you are correct, we do not need those dependencies in there. I believe I had them in there as I was thinking initially of a heavier integration test - however ended up just verifying the expected properties were set etc. I forgot to remove the deps once I switched gears on that. 

2. The security is kicking in due to `ManagementWebSecurityAutoConfiguration` being available on classpath now. If you look at the [RestDocsTestApplication](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/restdocs/RestDocsTestApplication.java#L27) used by the IT you will notice it was already excluding SecurityAutoConfiguration for this same reason. The test goes green when we include the `ManagementWebSecurityAutoConfiguration` in the exclude: 
```
@SpringBootApplication(exclude = { SecurityAutoConfiguration.class, ManagementWebSecurityAutoConfiguration.class })
```

**Comment by snicoll** [comment]
@bono007 thank you very much for making your first contribution to Spring Boot. I've polished your contribution in 
3530ac9.

**Comment by bono007** [comment]
Ha! Its not my first contribution to Spring Boot. But thank you for the warm welcome ;)

**Comment by wilkinsona** [comment]
This is your third first contribution no less 😬. Sorry about that. IIRC, GitHub kept listing you as a first-time contributor for a while. That's no longer the case so hopefully this will be your last first contribution.

**Comment by snicoll** [comment]
I know something was off when I saw the badge but I decided to go with the warm welcome anyway. Thank you for making your third first contribution, indeed 😂

**Comment by bono007** [comment]
3rd times a charm - so they say. ;)
